codice/ddf
# DDF-5360 Changed user ID based OAuth token storage to session ID to better handle the guest user.

#### What does this PR do?
Previously, the user's email address or username (if an email address is not present) was used to store users' OAuth tokens. This would cause problems for guest users. To solve that, we're switching to store tokens based on session IDs.

#### Who is reviewing it? 
@bakejeyner @SmithJosh 

#### Select relevant component teams: 
@codice/security 

#### Ask 2 committers to review/merge the PR and tag them here.
@garrettfreibott
@stustison 

#### How should this be tested?
Needs: https://github.com/codice/ddf-ui/pull/74
Create two sources (CSW and OpenSearch) to use OAuth (don't forget to add `/services=OIDC` to the webcontext policy manager and to configure the OIDC Handler)
- [ ] Both sources should be available
- [ ] Test querying the CSW source
    - Should get a notification to log in
    - Sign in and re-query
    - Query should succeed with no errors
- [ ] Test querying the Open Search source
    - Should get a notification to **authorize the source**
    - Authorize and re-query
    - Query should succeed with no errors
- [ ] Test session-based storage works
    - After logging in and querying successfully, sign out of DDF and Keycloak
    - Sign back in as the same user and try querying
    - You should be prompted to log in to Keycloak again 
- [ ] Test querying as guest
    - "Sign in" to intrigue as guest
    - Query a source and sign in to Keycloak
    - "Sign in" to intrigue as guest in another browser or in incognito mode
    - The second guest should be prompted to sign in before querying a source
- [ ] Test downloading one of the results you get back
- [ ] Test downloading an old result
    - After querying a source, remove your oauth tokens 
    - Try downloading one of the results you had
    - You should get redirected to sign in

#### What are the relevant tickets?
Fixes: #5360 

#### Checklist:

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.
